lib/craft/machine.ex

```diff
@@ -344,7 +379,30 @@ defmodule Craft.Machine do
   def handle_call({:query, :eventual, query}, from, state) do
     Logger.debug("executing query", logger_metadata(trace: {:query, :eventual, :quorum_read, from, query}))
 
-    {:reply, state.module.handle_query(query, state.private), state}
+    case state.module.handle_query(query, {:direct, from}, state.private) do
+      {:reply, reply} -> {:reply, reply, state}
+      :noreply -> {:noreply, state}
+    end
+  end
+
+  def handle_call({{:query_reply, reply}, query_from}, _from, state) do
+    is_pending? = MapSet.member?(state.pending_parallel_queries, query_from)
+    quorum_happend? = !is_nil(state.last_quorum_at) and :erlang.monotonic_time(:millisecond) > state.last_quorum_at
+
+    cond do
+      is_pending? and quorum_happend? ->
+        GenServer.reply(query_from, reply)
+        {:reply, :ok, %{state | pending_parallel_queries: MapSet.delete(state.pending_parallel_queries, query_from)}}
+      is_pending? ->
+        {:reply, :ok,
+         %{
+           state
+           | pending_parallel_queries: MapSet.delete(state.pending_parallel_queries, query_from),
+             client_query_results: [{query_from, reply} | state.client_query_results]
+         }}
+      true ->
+        {:reply, :noop, state}
+    end
```

Comment:
```suggestion
    if is_pending? do
      if quorum_happend? do
        GenServer.reply(query_from, reply)
    
        {:reply, :ok, %{state | pending_parallel_queries: MapSet.delete(state.pending_parallel_queries, query_from)}}
      else
        {:reply, :ok,
         %{
           state
           | pending_parallel_queries: MapSet.delete(state.pending_parallel_queries, query_from),
             client_query_results: [{query_from, reply} | state.client_query_results]
         }}
      end
    else
      {:reply, :noop, state}
    end
```

conds with repeated clause components are clearer (to me) as `if`s